### PR TITLE
[Azure Maps] Async Batch API schema change

### PR DESCRIPTION
**Pull request by chcmsft**
## Change
- change the AsyncBatchManagement API version from 2024-04-01-preview to 2024-10-01-preview
- change the Search API version from 2024-04-01-preview to 2024-10-01-preview
- add new Route API version 2024-10-01-preview
- remove Route API version 2024-04-01-preview
- remove route directions async batch api from both 2024-05-01-preview and 2024-06-01 preview api version
- object schema change of AsyncBatchOperation object in AsyncBatchManagement API
- request body change of async batch APIs in both Search 2024-10-01-preview and Route 2024-10-01-preview

### API version change
We change the api version of AsyncBatchManagement/Route/Search from 2024-04-01-preview to 2024-10-01-preview because of the release plan. The target is to release the async batch APIs with api version 2024-10-01-preview. Thus, Route 2024-05-01-preview and 2024-06-01-preview don't contain the async batch route direction API anymore and need to remove the async batch route directions api from both 2024-05-01-preview and 2024-06-01-preview api version.

#### Request body of geocoding async batch, reverse geocoding async batch, route directions async batch

We change the design from processing one input blob and generate one output blob to processing multiple input blobs and transform them into multiple output blobs.
Thus, the `inputBlobUrl` is replaced by `inputFolderPath`. the `outputStorageUrl` is replaced by `outputFolderPath`.

##### original request body
```
{
    "inputBlobUrl": "https://exampleinputstorage.blob.core.windows.net/asyncbatch/reverseGeocodingBatchItems.json",
    "outputStorageUrl": "https://exampleoutputstorage.blob.core.windows.net/",
    "msiClientId": "12345678-abcd-1234-efgh-12345678ijkl"
}
```

##### new request body
```
{
    "inputFolderPath": "https://exampleinputstorage.blob.core.windows.net/asyncbatch/input/",
    "outputFolderPath": "https://exampleoutputstorage.blob.core.windows.net/asyncbatch/output/",
    "msiClientId": "12345678-abcd-1234-efgh-12345678ijkl"
}
```

#### AsyncBatchManagement Get Operation API response schema change

due to the requirement change from processing single blob to multiple blobs, I tweak the operation status response according to the request body change above.

##### original
```
{
    "operationId": "123e4567-e89b-12d3-a456-426614174000",
    "operationType": "Geocode",
    "inputBlobUrl": "https://examplestorage.blob.core.windows.net/asyncbatch/geocodingBatchItems.json",
    "outputStorageUrl": "https://exampleoutputstorage.blob.core.windows.net/",
    "msiClientId": "87654321-abcd-1234-efgh-12345678ijkl",
    "createdDateTime": "2023-01-01T00:00:00Z",
    "lastActionDateTime": "2023-01-01T00:00:00Z",
    "percentComplete": 0,
    "totalItemCount": 1000,
    "status": "NotStarted"
}
```

##### new
```
{
    "id": "123e4567-e89b-12d3-a456-426614174000",
    "kind": "Geocode",
    "inputFolderPath": "https://examplestorage.blob.core.windows.net/asyncbatch/input/",
    "outputFolderPath": "https://exampleoutputstorage.blob.core.windows.net/asyncbatch/output/",
    "msiClientId": "87654321-abcd-1234-efgh-12345678ijkl",
    "createdDateTime": "2023-01-01T00:00:00Z",
    "lastActionDateTime": "2023-01-01T00:00:00Z",
    "progress": "0/10",
    "status": "NotStarted"
}
```

### Note
These APIs are not published yet. Thus, no breaking change concern.

<a href="?expand=1&template=data_plane_template.md">Click here</a> to open a PR for a Data Plane API.

### CODE REVIEW

File: specification/maps/data-plane/AsyncBatchManagement/preview/2024-04-01-preview/asyncBatchManagement.json [hunk 1]
```diff
@@ -66,8 +66,8 @@
       "required": true,
       "x-ms-parameter-location": "client"
     },
-    "OperationId": {
-      "name": "operationId",
+    "Id": {
+      "name": "id",
```
Comment (weidongxu-microsoft) [inline specification/maps/data-plane/AsyncBatchManagement/preview/2024-04-01-preview/asyncBatchManagement.json:70]: nit, the name for path parameter does not affect REST API

**Review by weidongxu-microsoft — APPROVED** [review]
You need to pass breaking changes review.

**Comment by chcmsft** [comment]
@mikekistler @JeffreyRichter 
Please help to review this PR.
The change is made due to the review comment from this [PR review](https://github.com/Azure/azure-rest-api-specs/pull/29153).

We want to make our API schema consistent by adopting kind and id to status monitor API.
This is not a breaking change since these APIs are not published yet.

File: specification/maps/data-plane/AsyncBatchManagement/preview/2024-04-01-preview/asyncBatchManagement.json (hunk 1 above)
Comment (mikekistler) [reply specification/maps/data-plane/AsyncBatchManagement/preview/2024-04-01-preview/asyncBatchManagement.json:70]: This change is not necessary for compliance with Azure REST API Guidelines. You could leave the parameter name as `operationId` and that would be fine. It's also fine to change it to `id`, so up to you whether you keep this change or revert it.

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

**Comment by chcmsft** [comment]
@mikekistler 
Thanks for your approval! If it's all good to you, can you help to remove the breakingChangedReviewRequired tag and merge this PR?

**Comment by chcmsft** [comment]
@weidongxu-microsoft 
The API schema is changed due to spec change. Can you help to review the new changes?

**Comment by weidongxu-microsoft** [comment]
Could you summarize that's changed after last review/approval?

Which commit ID need to be reviewed? Is there major API changes? etc.

**Comment by chcmsft** [comment]
@weidongxu-microsoft I put major changes in the PR description. 
https://github.com/Azure/azure-rest-api-specs/pull/29284#issue-2326934509

Please start the review from the commit e1fbfba88287352e231116571571c90f9c3e83cd.

**Comment by weidongxu-microsoft** [comment]
One major question is that the PR contains some move of Swagger or remove of API from existing Swagger.

Typically, one should not remove an api-version, unless you are discontinuing it (but this better to be done in a separate PR), and one should not remove API from existing api-version (you can discontinue a preview api-version, but I don't see a reason to remove 1 API from an api-version).

**Comment by chcmsft** [comment]
@mikekistler
FYI I adjusted the api version of the swagger.
I listed the reason in the description of the PR. Please refer to it.

**Comment by chcmsft** [comment]
@weidongxu-microsoft for the change other than api version, do you have any advice?

**Review by weidongxu-microsoft — COMMENTED** [review]
I don't have major concern about the API changes.

A few questions though.

File: specification/maps/data-plane/AsyncBatchManagement/preview/2024-10-01-preview/asyncBatchManagement.json [hunk 2]
```diff
@@ -225,16 +211,21 @@
...
+        "outputFolderPath": {
+          "description": "The URL that points to a Azure blob storage folder where the batch results will be stored.",
           "type": "string",
           "format": "uri"
         },
+        "allowBlobOverwrite": {
+          "description": "If a blob with the same name as the batch result already exists in the output folder, the transformation will terminate to avoid overwriting the existing blob. If the allowBlobOverwrite option is set to true, the existing blob will be overwritten.",
+          "type": "boolean",
+          "default": false
+        },
```
[the remark this answers is not in the document]
Comment (chcmsft) [reply specification/maps/data-plane/AsyncBatchManagement/preview/2024-10-01-preview/asyncBatchManagement.json:228]: yes, it's only used in response. default property removed.

File: specification/maps/data-plane/AsyncBatchManagement/preview/2024-10-01-preview/asyncBatchManagement.json [hunk 3]
```diff
@@ -249,23 +240,17 @@
...
-          "format": "int32",
-          "description": "Percentage indicating the completion status of the operation."
-        },
-        "totalItemCount": {
-          "type": "integer",
-          "format": "int32",
-          "description": "Total number of items in the batch operation."
+        "progress": {
+          "type": "string",
+          "description": "X out of Y indicating how many input blobs have been transformed."
```
[the remark this answers is not in the document]
Comment (chcmsft) [reply specification/maps/data-plane/AsyncBatchManagement/preview/2024-10-01-preview/asyncBatchManagement.json]: update accordingly

File: specification/maps/data-plane/AsyncBatchManagement/preview/2024-10-01-preview/asyncBatchManagement.json [hunk 4]
```diff
@@ -225,16 +211,21 @@
...
           "type": "string",
           "format": "uri"
         },
-        "outputStorageUrl": {
-          "description": "The URL that points to a blob storage where the batch results will be stored.",
+        "outputFolderPath": {
+          "description": "The URL that points to a Azure blob storage folder where the batch results will be stored.",
           "type": "string",
           "format": "uri"
         },
```
[the remark this answers is not in the document]
Comment (chcmsft) [reply specification/maps/data-plane/AsyncBatchManagement/preview/2024-10-01-preview/asyncBatchManagement.json:223]: Service will read all blobs in the folder and get the result of them one by one then store the output to the specified outputFolderPath.

Let's say
A customer provides an input folder path, "https://teststorage.blob.core.windows.net/input", there are 3 blobs in the folder
- input_1.json
- input_2.json
- input_3.json
The customer also provides an output folder path, "https://teststorage.blob.core.windows.net/output".
service will read input_1.json, input_2.json, input_3.json and create 3 results files input_1_res.json, input_2_res.json, input_3_res.json to the output folder path.

File: specification/maps/data-plane/Search/preview/2024-10-01-preview/search.json [hunk 5]
```diff
@@ -939,24 +939,29 @@
       "type": "object",
       "description": "Universal asynchronous batch request body.",
       "properties": {
-        "inputBlobUrl": {
-          "description": "The URL to the input blob containing the batch of queries. This should be a valid URL pointing to a blob of an Azure Storage Account",
+        "inputFolderPath": {
```
Comment (pbrasil) [inline specification/maps/data-plane/Search/preview/2024-10-01-preview/search.json:942]: Use same text and var name as suggested on "Batch Management" in this PR.

File: specification/maps/data-plane/Search/preview/2024-10-01-preview/search.json [hunk 6]
```diff
@@ -939,24 +939,29 @@
...
+          "description": "The URL that points to a Azure blob storage folder where the batch input blobs will be stored.",
           "type": "string",
           "format": "uri"
         },
-        "outputStorageUrl": {
-          "description": "The URL that points to a blob storage where the batch results will be stored.",
+        "outputFolderPath": {
```
Comment (pbrasil) [inline specification/maps/data-plane/Search/preview/2024-10-01-preview/search.json:947]: Use same text and var name as suggested on "Batch Management" in this PR.

File: specification/maps/data-plane/Search/preview/2024-10-01-preview/search.json [hunk 7]
```diff
@@ -939,24 +939,29 @@
...
+          "description": "The URL that points to a Azure blob storage folder where the batch results will be stored.",
           "type": "string",
           "format": "uri"
         },
+        "allowBlobOverwrite": {
```
Comment (pbrasil) [inline specification/maps/data-plane/Search/preview/2024-10-01-preview/search.json:952]: Use same text as suggested on "Batch Management" in this PR.

**Review by pbrasil — CHANGES_REQUESTED** [review]
First wave of doc updates. Mostly naming and consistency updates.

**Comment by winest** [comment]
Changes:
1. Change api-version from 2024-10-01-preview to 2024-07-01-preview to align with all azure maps service in this semester
2. Fix naming of inputFolderPath/outputFolderPath to inputContainerUri/outputContainerUri based on review comment
3. Fix descriptions based on review comment
4. Rebase on latest Azure/main branch
